### PR TITLE
Add Matomo cookie banner

### PR DESCRIPTION
**Pull request by ruester**
# Add cookie banner for Matomo tracking

Fixes https://github.com/hifis-net/RSD-as-a-service/issues/37
See also #392

See [Matomo tracking-consent documentation](https://developer.matomo.org/guides/tracking-consent) for technical details.

Changes proposed in this pull request:

* add a banner on all pages to ask the user for consent on being tracked (+cookies) from Matomo
* for this we use the library [react-cookie-consent](https://www.npmjs.com/package/react-cookie-consent) which offers a solution for a customizable cookie banner
* on a separate page the user is able to revoke or accept the tracking consent again

How to test:

* use a private browser tab or remove all your cookies for localhost
* opening any page of the RSD will show an overlay asking the user to consent/decline to the tracking+cookies of Matomo
* no requests on first page load are send out to Matomo
* accepting the tracking lead to requests being send out to the Matomo server
* by clicking on "decline" no requests are send to the Matomo ever and until this cookie is valid
* at the page `/cookies` the user gets some information about cookies in general and is able to revoke the matomo tracking consent again, each of them having the effect of stopping or starting the tracking of Matomo

PR Checklist:

*   [ ] Increase version numbers in `docker-compose.yml`
*   [ ] Link to a GitHub issue
*   [ ] Update documentation
*   [ ] Tests

### CODE REVIEW

**Comment by ruester** [comment]
@dmijatovic This is what I have so far. Its quite a mess.. mostly due to the lack of my react/nextjs skills :sweat_smile: 

I also wanted to add a link in the footer to the cookie information site, but I will wait for a decision on how to integrate the "revoke feature" first (https://github.com/research-software-directory/RSD-as-a-service/issues/392#issuecomment-1203808123)

**Review by cmeessen — COMMENTED** [review]
Looks like it works, thank you. I get some warnings in the console:

* on the frontpage, no cookies exist yet
* Click "Reject"
* `There was an error setting cookie `mtm_consent_removed`. Please check domain and path.`
* Is this because of the mismatch between `localhost` and `MATOMO_URL`?
* now I open `localhost/cookies` and click on accept, which prints the following warning:
`
Cookie “cookie_consent” does not have a proper “SameSite” attribute value. Soon, cookies without the “SameSite” attribute or with an invalid value will be treated as “Lax”. This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the “SameSite=None“ attribute to it. To know more about the “SameSite“ attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite`

**Comment by ruester** [comment]
There is currently also a problem for environment variables in the production build. I cannot get it to work so that the function components have the value for `MATOMO_URL` and `MATOMO_ID`... Its strange that it works for `_document.tsx` but not for `_app.tsx` or `cookies.tsx`.
It seems like those variables get deleted with some step, here I recorded a strange behaviour when reloading the page a few times:

https://user-images.githubusercontent.com/1287885/182875888-1130814a-06aa-4ba6-9ee1-5f8637a49750.mov

All the documentations I could find state that it should work. I also tried `publicRuntimeConfig` and `NEXT_PUBLIC_...`

@dmijatovic Could you help me getting the environment variables `MATOMO_*` to work also in production build?

**Comment by ruester** [comment]
It seems to be a problem with the standalone mode for the production build: https://github.com/vercel/next.js/issues/38119
But if I use the latest version of next (12.2.4) then I run into problems with jest: https://github.com/aelbore/esbuild-jest/issues/26

I've now found a solution with using `publicRuntimeConfig` and without the standalone mode, but this should only be considered a temporary workaround and to prove that something is not working as expected (at least according to all the documentation about environment variables in nextjs/react I've read :wink:)

**Comment by dmijatovic** [comment]
@ruester I started looking at this issue. Is there any specific reason for adding 2 additional dependencies: react-cookie-consent and react-use-cookie? Maybe we could solve it with dependencies we already have material ui, cookie.js lib and custom js?

Concerning the env variables. Accessing env variables from functional components (frontend only) is limited for security reasons. NEXT_PUBLIC...variables are available at build time only which is not sufficient for this case. AFAIK we can access env variables in the following ways:
- using getServerSideProps function defined at the page level, which unfortunately is not supported in _app.tsx. 
- by making a request to node api as done with auth env variables
- by loading mounted settings.json (also an fetch request) as I did it in [themes PR](https://github.com/research-software-directory/RSD-as-a-service/pull/410).

I have no experience with the approach you tried, [publicRuntimeConfig](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) but the documentation also points out that this approach only works with the server side pages (when page is using getServerSideProps) as far as I understand it.

To sum up, I stared looking at this issue and will come back with suggested approach after I make some tests. I am not familiar with Matomo so I will need to look at the documentation. Maybe you can get me at speed? How Matomo knows that user has given the consent for monitoring? Does cookie need to have specific name/values in it?

**Comment by ruester** [comment]
> To sum up, I stared looking at this issue and will come back with suggested approach after I make some tests. I am not familiar with Matomo so I will need to look at the documentation. Maybe you can get me at speed? How Matomo knows that user has given the consent for monitoring? Does cookie need to have specific name/values in it?

Many thanks for having a look at this. I thinks the approach with the request to the node API is the way to go here, I did not thought of that :see_no_evil: 
I can go ahead, if you want, and change my current pull request to use the API to get the environment variable, but this may also slow down the first page load quite a bit?

Matomo sets specific cookies values so it knows if the user has given consent, see here: https://developer.matomo.org/guides/tracking-consent

You can test if Matomo is activated (or not) by reloading the page and then look at the developer console for a request to the given Matomo URL (or by looking at the cookie values in the developer console).

**Comment by ruester** [comment]
> @ruester I started looking at this issue. Is there any specific reason for adding 2 additional dependencies: react-cookie-consent and react-use-cookie? Maybe we could solve it with dependencies we already have material ui, cookie.js lib and custom js?

Those may be left overs from my testing.. I think we do not need the react-use-cookie library for now :+1:
But the react-cookie-consent library was quite helpful since the cookie banner was easy to set up with this...

**Comment by ruester** [comment]
Ah sorry.. just saw the new PR now (https://github.com/research-software-directory/RSD-as-a-service/pull/488)
